core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -654,31 +653,29 @@ private[zio] final class FiberContext[E, A](
     }
 
   @tailrec
-  private[this] final def enterAsync(): Boolean = {
+  private[this] final def enterAsync(zio: ZIO[_, _, _]): Boolean = {
     val oldState = state.get
 
     oldState match {
       case Executing(_, observers, interrupt) =>
-        val newState = Executing(FiberStatus.Suspended(interruptible), observers, interrupt)
+        val interruptible = this.interruptible
+        val newState      = Executing(FiberStatus.Suspended(interruptible, zio), observers, interrupt)
 
-        if (!state.compareAndSet(oldState, newState)) enterAsync()
-        else if (shouldInterrupt) {
-          // Fiber interrupted, so go back into running state:
-          exitAsync()
-          false
-        } else true
+        if (interruptible && interrupt) false
```

Comment:
This feels odd... What happens when this fiber is interrupted
```scala
ZIO.never.ensuring(ZIO.sleep(5.seconds).interruptible *> putStrLn("Waited 5s"))
```
 ? Does it wait 5 seconds? Did it wait 5 seconds with previous code (I think no)? Could be that we had another bug in this logic... @jdegoes

Reply:
Hi, I checked this and yep the behavior is incorrect in ZIO 1.0.0-RC15 – "Waited 5s" is never printed - https://github.com/zio/zio/issues/2005

Reply:
Guess this should be ignored for now in this PR